File: test/utils/sla_test.dart
import 'package:civic_connect/models/issue.dart';
import 'package:civic_connect/utils/sla.dart';
import 'package:flutter_test/flutter_test.dart';

Issue _issue({
  required String category,
  required DateTime createdAt,
  String status = 'Pending',
}) {
  return Issue(
    id: '66bc2e2a0f8b1c4d4b123457',
    userId: 'u1',
    title: 'Test complaint',
    category: category,
    imageUrl: 'http://example.test/a.jpg',
    latitude: 23.03,
    longitude: 72.58,
    timestamp: createdAt,
    createdAt: createdAt,
    status: status,
  );
}

void main() {
  // Pinned so the tests do not drift with the wall clock.
  final now = DateTime(2026, 8, 20, 12);

  group('response windows', () {
    test('urgent categories get a short window', () {
      expect(SlaPolicy.windowFor('water'), 2);
      expect(SlaPolicy.windowFor('electricity'), 2);
    });

    test('wear-and-tear categories get a long window', () {
      expect(SlaPolicy.windowFor('road'), 10);
      expect(SlaPolicy.windowFor('pothole'), 7);
    });

    test('an unknown category falls back rather than throwing', () {
      expect(SlaPolicy.windowFor('meteor_strike'), 7);
    });

    test('category matching ignores case', () {
      expect(SlaPolicy.windowFor('WATER'), 2);
    });
  });

  group('evaluate', () {
    test('reports days remaining inside the window', () {
      final issue = _issue(
        category: 'pothole', // 7 day window
        createdAt: now.subtract(const Duration(days: 5)),
      );

      final sla = SlaPolicy.evaluate(issue, now: now);

      expect(sla.state, SlaState.onTrack);
      expect(sla.days, 2);
      expect(sla.label, 'Due in 2 days');
    });

    test('flags the last day before the deadline', () {
      final issue = _issue(
        category: 'pothole',
        createdAt: now.subtract(const Duration(days: 6)),
      );

      final sla = SlaPolicy.evaluate(issue, now: now);

      expect(sla.state, SlaState.dueSoon);
      expect(sla.label, 'Due tomorrow');
    });

    test('flags work past its deadline and counts the days late', () {
      final issue = _issue(
        category: 'water', // 2 day window
        createdAt: now.subtract(const Duration(days: 5)),
      );

      final sla = SlaPolicy.evaluate(issue, now: now);

      expect(sla.state, SlaState.overdue);
      expect(sla.isOverdue, isTrue);
      expect(sla.days, 3);
      expect(sla.label, 'Overdue by 3 days');
    });

    test('singular day reads correctly', () {
      final issue = _issue(
        category: 'water',
        createdAt: now.subtract(const Duration(days: 3)),
      );

      expect(SlaPolicy.evaluate(issue, now: now).label, 'Overdue by 1 day');
    });

    test('stops the clock once the complaint is resolved', () {
      final issue = _issue(
        category: 'water',
        createdAt: now.subtract(const Duration(days: 90)),
        status: 'Resolved',
      );

      final sla = SlaPolicy.evaluate(issue, now: now);

      expect(sla.state, SlaState.closed);
      expect(sla.isClosed, isTrue);
      expect(sla.isOverdue, isFalse);
    });

    test('stops the clock once the complaint is rejected', () {
      final issue = _issue(
        category: 'water',
        createdAt: now.subtract(const Duration(days: 90)),
        status: 'Rejected',
      );

      expect(SlaPolicy.evaluate(issue, now: now).isClosed, isTrue);
    });
  });
}

File: server/test/sla.test.js
const test = require('node:test');
const assert = require('node:assert/strict');
const sla = require('../config/sla');

const DAY = 86400000;
const now = new Date('2026-08-20T12:00:00Z');

const issue = ({ category = 'pothole', status = 'Pending', ageDays = 0 }) => ({
  category,
  status,
  createdAt: new Date(now.getTime() - ageDays * DAY),
});

test('urgent categories get a short response window', () => {
  assert.equal(sla.windowFor('water'), 2);
  assert.equal(sla.windowFor('electricity'), 2);
});

test('wear-and-tear categories get a long response window', () => {
  assert.equal(sla.windowFor('road'), 10);
  assert.equal(sla.windowFor('pothole'), 7);
});

test('an unknown category falls back instead of producing NaN', () => {
  assert.equal(sla.windowFor('meteor_strike'), sla.DEFAULT_WINDOW_DAYS);
  assert.equal(sla.windowFor(undefined), sla.DEFAULT_WINDOW_DAYS);
  assert.equal(sla.windowFor(null), sla.DEFAULT_WINDOW_DAYS);
});

test('category matching ignores case', () => {
  assert.equal(sla.windowFor('WATER'), 2);
});

test('the deadline is the filing date plus the window', () => {
  const due = sla.dueDate(issue({ category: 'water', ageDays: 0 }));
  assert.equal(due.getTime(), now.getTime() + 2 * DAY);
});

test('work inside its window is not overdue', () => {
  assert.equal(sla.isOverdue(issue({ category: 'pothole', ageDays: 3 }), now), false);
});

test('work past its window is overdue', () => {
  assert.equal(sla.isOverdue(issue({ category: 'water', ageDays: 5 }), now), true);
});

test('closed work is never overdue, however old', () => {
  const ancient = { category: 'water', ageDays: 900 };
  assert.equal(
    sla.isOverdue(issue({ ...ancient, status: 'Resolved' }), now),
    false
  );
  assert.equal(
    sla.isOverdue(issue({ ...ancient, status: 'Rejected' }), now),
    false
  );
});

test('isClosed recognises exactly the terminal states', () => {
  assert.equal(sla.isClosed({ status: 'Resolved' }), true);
  assert.equal(sla.isClosed({ status: 'Rejected' }), true);
  assert.equal(sla.isClosed({ status: 'Pending' }), false);
  assert.equal(sla.isClosed({ status: 'In Progress' }), false);
});

test('the policy table matches the client fallback in lib/utils/sla.dart', () => {
  // Guards the one duplication in the codebase. If this fails, the citizen app
  // and the municipal dashboard disagree about what is late.
  assert.deepEqual(sla.WINDOW_DAYS, {
    water: 2,
    electricity: 2,
    street_light: 3,
    garbage: 3,
    drainage: 5,
    pothole: 7,
    road: 10,
    other: 7,
  });
});
